src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,135 @@
+
+  // Size of this cookie in bytes
+  def size: Int
+}
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: UcxMemoryBlock
+
+  // Called when the transport is done with the Memory, so we can unmap it for example
+  // the transport is not to use this Memory anymore
+  def doneWithMemory(mem: UcxMemoryBlock)
+}
+
+trait OprationStatus extends Enumeration {
+  val SUCCESS = Value(0)
+  val FAILURE = Value(1)
+}
+
+/**
+ * Operation statistic, like completionTime, transport used, protocol used, etc.
+ */
+trait OperationStats
+
+trait TransportError extends Throwable
+
+trait OperationResult {
+  def getStatus: OprationStatus
+  def getError: TransportError
+  def getStats: OperationStats
+}
+
+/**
+ * Async operation callbacks
+ */
+trait OperationCallback {
+  def onComplete(result: OperationResult)
+}
+
+/**
+ * Transport flow example:
+ * val transport = new UcxShuffleTransport()
+ * transport.init()
+ *
+ * Mapper/writer:
+ * val cookies = transport.register(dataBlockIds, blocks)
+ * val metadataBlockIds = dataBlockIds.map(dataBlockId => new MetadataBlockId(dataBlockId))
+ * cookies.foreach(cookie => cookie.writeToMemory(metadataBlocks))
+ * transport.register(metadataBlockIds, metadataBlocks) // we don't care for metadata cookies
+ *
+ *
+ * Reducer:
+ * 1. First need to fetch metadata for blockIds:
+ * transport.fetchBlocksByBlockIds(remoteExecutor, metadataBlockIds, resultBuffer, callback)
+ * 2. Deserialize cookies from result buffer:
+ * val cookies = blockIds.map(_ => transport.getCookieFromMemory(resultBuffer))
+ * 3. Fetch blocks by cookies:
+ * transport.fetchBlocksByCookies(remoteExecutor, dataBlockIds, cookies, resultBuffer, callback)
+ * 4. Progress communications:
+ * while(noMoreBlocks) { transport.progress() }
+ *
+ *
+ * transport.unregister(blockIds)
+ * transport.close()
+ */
+trait UcxShuffleTransport {
+
+  /**
+   * Initialize transport resources. This function should get called after ensuring that SparkConf
+   * has the correct configurations since it will use the spark configuration to configure itself.
+   */
+  def init()
+
+  /**
+   * Close all transport resources
+   */
+  def close()
+
+  /**
+   * Registers blocks using blockId on SERVER side.
+   */
+  def register(blockIds: Seq[BlockId], blocks: Seq[Block]): Seq[Cookie]
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  def mutate(blockId: BlockId, block: Block, callback: OperationCallback)
+
+  /**
+   * Indicate that this blockId is not needed any more by an application
+   */
+  def unregister(blockIds: Seq[BlockId])
+
+  /**
+   * Fetch remote blocks by blockIds
+   */
+  def fetchBlocksByBlockIds(executorId: String, blockIds: Seq[BlockId],
```

Comment:
what is the relationship between `blockId` N and `resultBuffer`?

This is not obvious from the API, and as an application I'll need to know how the transport intends to put those bytes into `UcxMemoryBlock` in order to decode it. Right?

Reply:
I honestly prefer it be one-to-one. `requestBlockId(blockId, memory)`, `requestCookie(cookie, memory)`. Then it is more obvious that `memory` is the buffer corresponding to that block as the writer registered it.

Reply:
That said having the many-to-many one is nice, as it allows the transport to batch them to some degree:
`requestBlockIs(blockIds, memories)`, and `requestCookies(cookies, memories)`, where length of blockId/cookie and `memories` is the same, and so it is still obvious what is at each `memory`.

Reply:
So for me, I am currently assuming the last one (many-to-many) exists.

Reply:
Transport will put N blockIds result contiguously. E.g.:
```
transport.register(block1, ByteBuffer.allocateDirect(1))
transport.register(block2, ByteBuffer.allocateDirect(2)
transport.register(block3, ByteBuffer.allocateDirect(3)

transport.fetchBlocksByBlockIds(Array(block1, block2, block3), ByteBuffer.allocateDirect(5))
// | 1 byte block1| 2 bytesBlock2 | 3 bytes block3|
```

Reply:
For the cookie case this could work (since the size is known to the client), but not for `fetchBlocksByBlockIds` as you have in your example. The client doesn't know the size of each block on its side so it can't know the layout, unless the output buffer has a header with the layout of the output bytes in memory.

For the fetch by cookie case it also implies that you have to wait for all blocks to call the callback. This api doesn't allow for that case either ... I can't ask for all the blocks I want transferred and start consuming them as they arrive, instead the transport needs to wait for all blocks to show up in order to hand back the layout proposed.

We may want to allow the callback to have a start/onBlock/done semantics so the transport doesn't need to queue up memory, and it would clear up what-is-what in terms of the layout of the output buffer. Thoughts?

Reply:
> We may want to allow the callback to have a start/onBlock/done semantics so the transport doesn't need to queue up memory, and it would clear up what-is-what in terms of the layout of the output buffer. Thoughts?

RDMA doesn't allow you to have partial data transfer. It's all or none. For send/recv there could be some partial data with staging protocols, but it's certainly not - per block, and doesn't expose even in ucp.

Reply:
So it's up to the application to do load balancing.

Reply:
Yes, result memory should have headroom for `numBlocks * 4`, so first `numBlocks * 4` would be sizes of blocks

Reply:
I am not suggesting partial transfers. What I am suggesting is that the app requests N pieces of memory, and so the app needs to be able to tell where that memory ended up, if not in N output buffers. Otherwise the transport decides a layout for a buffer we have no access to, and the transport is **waiting** for all N blocks to show up in the buffer.

If the implementation in this case is that N block requests become 1 RDMA READ, then that's a bit better from the performance perspective. The sender needs to produce a block and put it somewhere, and then the client can fetch that memory. Is that what you are suggesting?

> Yes, result memory should have headroom for numBlocks * 4, so first numBlocks * 4 would be sizes of blocks

This is not going to work for the GPU. The header would go to the GPU as well, so we need to copy it back to the host, and slice up the device buffer, when this is really data only required in the CPU.

Can we not write anything to the destination buffer than what was written there by the register function? 

If you want to allow for this type of semantics, perhaps there should be a block layout structure that is host based, or we should reconsider this plan. I realize I am proposing a second RPC call.

Reply:
> If the implementation in this case is that N block requests become 1 RDMA READ, then that's a bit better from the performance perspective. The sender needs to produce a block and put it somewhere, and then the client can fetch that memory. Is that what you are suggesting?

Actually knowing the answer to this is key. The transport doesn't allocate memory, so it could be an issue, I'd like to know how that would implemented.

Reply:
So thinking about this more: I don't think we are setup to use a multiple blockId -> single contiguous memory transfer as of today. Main thing is we track memory per block in general, so blocks are consumed by downstream tasks. 

I think we could adapt, but I need to understand how this would implemented without creating more memory for example. Is it IOV?

And also, the other thing is that if you have a single block, will the header not be there?

Seems we are overloading this function a lot.. I'd say we don't do this unless we have a compelling case for it, and if so maybe add another function that means we need to have special block tracking in the application side so we can free up memory accordingly, deal with offsets potentially, etc.

Reply:
This is for CPU shuffle metadata blocks. We know *exact* cookie size. So we can request 10 metadata blocks to 1 contiguous recv buffer. It's not IOV. But basically 1 buffer send/recv with multiple blocksIds.

Reply:
You can always use this API for a single block `fetchBlocksByBlockIds(Array(block))`. Then only the first 4 bytes would be the actual size of the block.

Reply:
The block returned should be exactly the same block registered. No extra bytes. So this is not going to work for us.

Otherwise something needs to copy a buffer every time write or read it from the transport layer, and that is not going to work.

Reply:
OK, so let's assume: if you know exact blocks sizes in advance you can submit many - to one memory buffer request.  If no - just submit a single block request. `OperationCallback` now has an optional field `def recvSize: Long` - it's not guaranteed you'll get it (if it's one-sided operation).

Reply:
What is recvSize when it is not valid? And in what cases is it set? At the minimum it should be something that is a scala Option, but we need to document clearly why it exists.

Reply:
Also recvSize doesn’t seem to address the many-to-one question right? Unless we are back to the buffer received has a layout that is transport specific and the app needs to decipher or the recvSize is per block included in the larger allocation.

Reply:
`recvSize` should be always correct. If it's one-sided communication it'll always be : `cookies.map(cookie => cookie.length).sum`. For `fetchBlocksByBlockIds` it'll tell total recv bytes. It'll not solve many-to-one problem if you don't know blocks length.  But it'll help at least for a single block if you don't know it's size

Reply:
Ok it is definitely good if recvSize is always defined (originally it seemed not well defined so that was confusing). For the many-to-one case what will the gpu impl do exactly? Or is that not a valid API call?